### PR TITLE
[3279] Restore support for creating a domain diagram automatically

### PR DESCRIPTION
**Pull request by sbegaudeau**
- [3279] Restore support for creating a domain diagram automatically

# Pull request template

## General purpose
What is the main goal of this pull request?
- [ ] Bug fixes
- [ ] New features
- [ ] Documentation
- [ ] Cleanup
- [ ] Tests
- [ ] Build / releng

## Project management
- [ ] Has the pull request been added to the relevant project and milestone? (Only if you know that your work is part of a specific iteration such as the current one)
- [ ] Have the `priority:` and `pr:` labels been added to the pull request? (In case of doubt, start with the labels `priority: low` and `pr: to review later`)
- [ ] Have the relevant issues been added to the pull request?
- [ ] Have the relevant labels been added to the issues? (`area:`, `difficulty:`, `type:`)
- [ ] Have the relevant issues been added to the same project and milestone as the pull request?
- [ ] Has the `CHANGELOG.adoc` been updated to reference the relevant issues?
- [ ] Have the relevant API breaks been described in the `CHANGELOG.adoc`? (Including changes in the GraphQL API)
- [ ] In case of a change with a visual impact, are there any screenshots in the `CHANGELOG.adoc`? For example in `doc/screenshots/2022.5.0-my-new-feature.png`

## Architectural decision records (ADR)
- [ ] Does the title of the commit contributing the ADR start with `[doc]`?
- [ ] Are the ADRs mentioned in the relevant section of the  `CHANGELOG.adoc`?

## Dependencies
- [ ] Are the new / upgraded dependencies mentioned in the relevant section of the `CHANGELOG.adoc`?
- [ ] Are the new dependencies justified in the `CHANGELOG.adoc`?

## Frontend

This section is not relevant if your contribution does not come with changes to the frontend.

### General purpose
- [ ] Is the code properly tested? (Plain old JavaScript tests for business code and tests based on React Testing Library for the components)

### Typing
We need to improve the typing of our code, as such, we require every contribution to come with proper TypeScript typing for both changes contributing new files and those modifying existing files.
Please ensure that the following statements are true for each file created or modified (this may require you to improve code outside of your contribution).

- [ ] Variables have a proper type
- [ ] Functions’ arguments have a proper type
- [ ] Functions’ return type are specified
- Hooks are properly typed:
	- [ ] `useMutation<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useQuery<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useSubscription<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useMachine<CONTEXT_TYPE, EVENTS_TYPE>(…)`
	- [ ] `useState<STATE_TYPE>(…)`
- [ ] All components have a proper typing for their props
- [ ] No useless optional chaining with `?.` (if the GraphQL API specifies that a field cannot be `null`, do not treat it has potentially `null` for example)
- [ ] Nullable values have a proper type (for example `let diagram: Diagram | null = null;`)

## Backend

This section is not relevant if your contribution does not come with changes to the backend.

### General purpose
- [ ] Are all the event handlers tested?
- [ ] Are the event processor tested?
- [ ] Is the business code (services) tested?
- [ ] Are diagram layout changes tested?

### Architecture
- [ ] Are data structure classes properly separated from behavioral classes?
- [ ] Are all the relevant fields final?
- [ ] Is any data structure mutable? If so, please write a comment indicating why
- [ ] Are behavioral classes either stateless or side effect free?

### How to test this PR?
_Please describe here the various use cases to test this pull request_

- [ ] Has the Kiwi TCMS test suite been updated with tests for this contribution?

### CODE REVIEW

File: packages/view/backend/sirius-components-view-diagram-customnodes-edit/src/main/java/org/eclipse/sirius/components/view/diagram/customnodes/provider/CustomNodesEditPlugin.java [hunk 1]
```diff
@@ -95,8 +63,7 @@ protected Object doGetImage(String key) throws IOException {
      * @return the singleton instance.
      * @generated
      */
-    @Override
-    public ResourceLocator getPluginResourceLocator() {
+    public static Implementation getPlugin() {
```
Comment (pcdavid) [inline packages/view/backend/sirius-components-view-diagram-customnodes-edit/src/main/java/org/eclipse/sirius/components/view/diagram/customnodes/provider/CustomNodesEditPlugin.java]: Aren't we losing the custom code needed to support SVG icons here?

File: packages/view/backend/sirius-components-view-diagram-customnodes-edit/src/main/java/org/eclipse/sirius/components/view/diagram/customnodes/provider/EllipseNodeStyleDescriptionItemProvider.java [hunk 2]
```diff
@@ -109,11 +109,11 @@ protected void addBorderLineStylePropertyDescriptor(Object object) {
     /**
      * This returns EllipseNodeStyleDescription.gif. <!-- begin-user-doc --> <!-- end-user-doc -->
      *
-     * @generated NOT
+     * @generated
      */
     @Override
     public Object getImage(Object object) {
-        return this.overlayImage(object, this.getResourceLocator().getImage("full/obj16/EllipseNodeStyle.svg"));
+        return this.overlayImage(object, this.getResourceLocator().getImage("full/obj16/EllipseNodeStyleDescription"));
```
Comment (pcdavid) [inline packages/view/backend/sirius-components-view-diagram-customnodes-edit/src/main/java/org/eclipse/sirius/components/view/diagram/customnodes/provider/EllipseNodeStyleDescriptionItemProvider.java]: `@generated NOT` + `EllipseNodeStyle.svg` lost.

File: packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/studio/services/representations/RelationEdgeDescriptionProvider.java [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+import org.eclipse.sirius.components.view.builder.generated.DiagramBuilders;
+import org.eclipse.sirius.components.view.builder.providers.IColorProvider;
+import org.eclipse.sirius.components.view.builder.providers.IEdgeDescriptionProvider;
+import org.eclipse.sirius.components.view.diagram.ArrowStyle;
+import org.eclipse.sirius.components.view.diagram.DiagramDescription;
+import org.eclipse.sirius.components.view.diagram.EdgeDescription;
+import org.eclipse.sirius.components.view.diagram.EdgeStyle;
+
+/**
+ * Used to create the inheritance edge description.
```
Comment (pcdavid) [inline packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/studio/services/representations/RelationEdgeDescriptionProvider.java]: Typo: "inheritance" => "relation"

File: packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/studio/services/representations/api/IDomainDiagramDescriptionProvider.java [hunk 4]
```diff
@@ -0,0 +1,22 @@
+/*******************************************************************************
+ * Copyright (c) 2024, 2024 Obeo.
```
Comment (pcdavid) [inline packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/studio/services/representations/api/IDomainDiagramDescriptionProvider.java]: 2024 twice

File: packages/sirius-web/backend/sirius-web/src/test/java/org/eclipse/sirius/web/application/controllers/diagrams/CustomNodesDiagramControllerTests.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    @BeforeEach
+    public void beforeEach() {
+        this.givenInitialServerState.initialize();
+    }
+
+    @Test
+    @DisplayName("Given a diagram with collapsed nodes by default, when it is opened, then some nodes are collapsed")
+    @Sql(scripts = {"/scripts/papaya.sql"}, executionPhase = Sql.ExecutionPhase.BEFORE_TEST_METHOD)
+    @Sql(scripts = {"/scripts/cleanup.sql"}, executionPhase = Sql.ExecutionPhase.AFTER_TEST_METHOD, config = @SqlConfig(transactionMode = SqlConfig.TransactionMode.ISOLATED))
+    public void givenDiagramWithCollapsedNodesByDefaultWhenItIsOpenedThenSomeNodesAreCollapsed() {
```
Comment (pcdavid) [inline packages/sirius-web/backend/sirius-web/src/test/java/org/eclipse/sirius/web/application/controllers/diagrams/CustomNodesDiagramControllerTests.java]: Test method name and @DisplayName do not match the actual content.

File: packages/sirius-web/backend/sirius-web/src/test/java/org/eclipse/sirius/web/services/diagrams/CustomNodesDiagramDescriptionProvider.java [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+    }
+
+    public String getRepresentationDescriptionId() {
+        return this.diagramIdProvider.getId(this.diagramDescription);
+    }
+
+    private View createView() {
+        ViewBuilder viewBuilder = new ViewBuilder();
+        View expandCollapseView = viewBuilder.build();
+        expandCollapseView.getDescriptions().add(this.createDiagramDescription());
```
Comment (pcdavid) [inline packages/sirius-web/backend/sirius-web/src/test/java/org/eclipse/sirius/web/services/diagrams/CustomNodesDiagramDescriptionProvider.java]: `expandCollapseView` does not match this particular view.

File: packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/studio/services/StudioProjectTemplateInitializer.java [hunk 7]
```diff
@@ -90,4 +117,12 @@ private Optional<ResourceSet> getResourceSet(IEditingContext editingContext) {
                 .map(IEMFEditingContext::getDomain)
                 .map(AdapterFactoryEditingDomain::getResourceSet);
     }
+
+    private Optional<DiagramDescription> findDomainDiagramDescription(IEditingContext editingContext) {
+        return this.representationDescriptionSearchService.findAll(editingContext).values().stream()
+                .filter(DiagramDescription.class::isInstance)
+                .map(DiagramDescription.class::cast)
+                .filter(diagramDescription -> Objects.equals("Domain", diagramDescription.getLabel()))
```
Comment (pcdavid) [inline packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/studio/services/StudioProjectTemplateInitializer.java]: This seems brittle to check on the label with a rather generic name; could we match on it's id and use a some random UUID to make sure we find the one we want?

**Review by pcdavid — CHANGES_REQUESTED** [review]

File: packages/view/backend/sirius-components-view-diagram-customnodes-edit/src/main/java/org/eclipse/sirius/components/view/diagram/customnodes/provider/CustomNodesEditPlugin.java (hunk 1 above)
Comment (sbegaudeau) [reply packages/view/backend/sirius-components-view-diagram-customnodes-edit/src/main/java/org/eclipse/sirius/components/view/diagram/customnodes/provider/CustomNodesEditPlugin.java]: It seems that we may have been missing a generated not somewhere. I'll restore the previous code

Comment (sbegaudeau) [reply packages/view/backend/sirius-components-view-diagram-customnodes-edit/src/main/java/org/eclipse/sirius/components/view/diagram/customnodes/provider/CustomNodesEditPlugin.java]: Done

File: packages/view/backend/sirius-components-view-diagram-customnodes-edit/src/main/java/org/eclipse/sirius/components/view/diagram/customnodes/provider/EllipseNodeStyleDescriptionItemProvider.java (hunk 2 above)
Comment (sbegaudeau) [reply packages/view/backend/sirius-components-view-diagram-customnodes-edit/src/main/java/org/eclipse/sirius/components/view/diagram/customnodes/provider/EllipseNodeStyleDescriptionItemProvider.java]: Done

File: packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/studio/services/representations/RelationEdgeDescriptionProvider.java (hunk 3 above)
Comment (sbegaudeau) [reply packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/studio/services/representations/RelationEdgeDescriptionProvider.java]: Done

File: packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/studio/services/representations/api/IDomainDiagramDescriptionProvider.java (hunk 4 above)
Comment (sbegaudeau) [reply packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/studio/services/representations/api/IDomainDiagramDescriptionProvider.java]: Done

File: packages/sirius-web/backend/sirius-web/src/test/java/org/eclipse/sirius/web/application/controllers/diagrams/CustomNodesDiagramControllerTests.java (hunk 5 above)
Comment (sbegaudeau) [reply packages/sirius-web/backend/sirius-web/src/test/java/org/eclipse/sirius/web/application/controllers/diagrams/CustomNodesDiagramControllerTests.java]: Done

File: packages/sirius-web/backend/sirius-web/src/test/java/org/eclipse/sirius/web/services/diagrams/CustomNodesDiagramDescriptionProvider.java (hunk 6 above)
Comment (sbegaudeau) [reply packages/sirius-web/backend/sirius-web/src/test/java/org/eclipse/sirius/web/services/diagrams/CustomNodesDiagramDescriptionProvider.java]: Done

File: packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/studio/services/StudioProjectTemplateInitializer.java (hunk 7 above)
Comment (sbegaudeau) [reply packages/sirius-web/backend/sirius-web-application/src/main/java/org/eclipse/sirius/web/application/studio/services/StudioProjectTemplateInitializer.java]: Given the recent improvements made in our tests, I think I now have what's needed for that. In the previous architecture it was only tested this way. I've updated the code.

**Review by pcdavid — APPROVED** [review]

**Pull request merged**